src/validation.cpp

```diff
@@ -2192,16 +2196,17 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
                 // data providers.
                 CScriptCheck check2(txdata.m_spent_outputs[i], tx, validation_cache.m_signature_cache, i,
                         flags & ~STANDARD_NOT_MANDATORY_VERIFY_FLAGS, cacheSigStore, &txdata);
-                if (check2())
-                    return state.Invalid(TxValidationResult::TX_NOT_STANDARD, strprintf("non-mandatory-script-verify-flag (%s)", ScriptErrorString(check.GetScriptError())));
-
-                // If the second check failed, it failed due to a mandatory script verification
-                // flag, but the first check might have failed on a non-mandatory script
-                // verification flag.
-                //
-                // Avoid reporting a mandatory script check failure with a non-mandatory error
-                // string by reporting the error from the second check.
-                error = check2.GetScriptError();
+                if (auto result2 = check2(); !result2.has_value()) {
```

Comment:
Is there perf reasons to put the `auto result2 = check();` into the `if` expression? Seems more readable to do

```c++
auto result2 = check2(); 
if (!result2.has_value()) {
  ...
}
```

IIUC I see why this is done on [`L2187`](https://github.com/bitcoin/bitcoin/pull/31112/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R2187) - to avoid doing the computation unnecessarily.